include/net/lwm2m.h

```diff
@@ -1000,21 +1020,57 @@ int lwm2m_engine_register_delete_callback(uint16_t obj_id,
 int lwm2m_engine_set_res_data(const char *pathstr, void *data_ptr, uint16_t data_len,
 			      uint8_t data_flags);
 
+/**
+ * @brief Update data size for a resource
+ *
+ * Use this function to set the new size of data in the buffer if you write
+ * to a buffer received by lwm2m_engine_get_res_buf().
+ *
+ * @param[in] pathstr LwM2M path string "obj/obj-inst/res(/res-inst)"
+ * @param[in] data_len Length of data
+ * @return 0 for success or negative in case of error.
+ */
+int lwm2m_engine_set_res_data_len(const char *pathstr, uint16_t data_len);
+
 /**
  * @brief Get data buffer for a resource
  *
  * Use this function to get the data buffer information for the specified LwM2M
  * resource.
  *
+ * If you directly write into the buffer, you must use lwm2m_engine_set_res_data_len()
+ * function to update the new size of the written data.
+ *
+ * All parameters except pathstr can NULL if you don't want to read those values.
+ *
+ * @param[in] pathstr LwM2M path string "obj/obj-inst/res(/res-inst)"
+ * @param[out] buffer_ptr Data buffer pointer
+ * @param[out] buffer_len Length of buffer
+ * @param[out] data_len Length of existing data in the buffer
+ * @param[out] data_flags Data buffer flags (such as read-only, etc)
+ *
+ * @return 0 for success or negative in case of error.
+ */
+int lwm2m_engine_get_res_buf(const char *pathstr, void **buffer_ptr, uint16_t *buffer_len,
+			     uint16_t *data_len, uint8_t *data_flags);
+
+/**
+ * @brief Get data buffer for a resource
+ *
+ * Use this function to get the data buffer information for the specified LwM2M
+ * resource.
+ *
+ * @deprecated Use lwm2m_engine_get_res_buf() as it can tell you the size of the buffer as well.
+ *
  * @param[in] pathstr LwM2M path string "obj/obj-inst/res(/res-inst)"
  * @param[out] data_ptr Data buffer pointer
- * @param[out] data_len Length of buffer
+ * @param[out] data_len Length of existing data in the buffer
  * @param[out] data_flags Data buffer flags (such as read-only, etc)
  *
  * @return 0 for success or negative in case of error.
  */
-int lwm2m_engine_get_res_data(const char *pathstr, void **data_ptr,
-			      uint16_t *data_len, uint8_t *data_flags);
+int lwm2m_engine_get_res_data(const char *pathstr, void **data_ptr, uint16_t *data_len,
```

Comment:
We should mark the function with `__deprecated` keyword, so that it produces a compiler warning.